File: 0/0/8/008.c
// Copyright (C) 2014 Jorge Aparicio

#include <stdio.h>

static int stride = 5;

int product(int* factors) {
  int out = 1;

  for (int i = 0; i < stride; i++) {
    out *= factors[i];
  }

  return out;
}

int main() {
  int chr, p, pos = 0, max = 0;
  int factors[5] = {0};

  FILE *file = fopen("008.in", "r");

  chr = fgetc(file);
  while (chr != EOF) {
    if (chr != '\n') {
      factors[pos] = chr - 48;
      p = product(factors);

      if (p > max) {
        max = p;
      }

      pos = (pos + 1) % stride;
    }

    chr = fgetc(file);
  }

  printf("%d\n", max);

  return 0;
}

File: 0/0/8/008.rs
// Copyright (C) 2014 Jorge Aparicio

use std::char::to_digit;
use std::io::BufferedReader;
use std::io::fs::File;

static stride: uint = 5;

fn main() {
    let (mut max, mut pos) = (0, 0);
    let mut factors: [uint, ..stride] = [0, ..stride];

    let mut content = BufferedReader::new(File::open(&Path::new("008.in")));

    for char in content.chars() {
        if char != '\n' {
            factors[pos] = to_digit(char, 10).unwrap();

            let prod = factors.iter().fold(1, |a, &b| a * b);

            if prod > max {
                max = prod;
            }

            pos = (pos + 1) % stride;
        }

    }

    println!("{}", max);
}

